pkg/storage/stores/shipper/downloads/index_set.go

```diff
@@ -0,0 +1,421 @@
+		boltdb, err := shipper_util.SafeOpenBoltdbFile(filePath)
+		if err != nil {
+			return err
+		}
+
+		t.dbs[fileName] = boltdb
+	}
+
+	for _, db := range toDelete {
+		err := t.cleanupDB(db)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// checkStorageForUpdates compares files from cache with storage and builds the list of files to be downloaded from storage and to be deleted from cache
+func (t *indexSet) checkStorageForUpdates(ctx context.Context, lock bool) (toDownload []storage.IndexFile, toDelete []string, err error) {
+	// listing tables from store
+	var files []storage.IndexFile
+
+	files, err = t.baseIndexSet.ListFiles(ctx, t.tableName, t.userID)
+	if err != nil {
+		return
+	}
+
+	listedDBs := make(map[string]struct{}, len(files))
+
+	if lock {
+		err = t.dbsMtx.rLock(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		defer t.dbsMtx.rUnlock()
+	}
+
+	for _, file := range files {
+		listedDBs[file.Name] = struct{}{}
+
+		// Checking whether file was already downloaded, if not, download it.
+		// We do not ever upload files in the object store with the same name but different contents so we do not consider downloading modified files again.
+		_, ok := t.dbs[file.Name]
+		if !ok {
+			toDownload = append(toDownload, file)
+		}
+	}
+
+	for db := range t.dbs {
+		if _, isOK := listedDBs[db]; !isOK {
+			toDelete = append(toDelete, db)
+		}
+	}
+
+	return
+}
+
+func (t *indexSet) AwaitReady(ctx context.Context) error {
+	return t.dbsMtx.awaitReady(ctx)
+}
+
+func (t *indexSet) downloadFileFromStorage(ctx context.Context, fileName, folderPathForTable string) error {
+	return shipper_util.DownloadFileFromStorage(filepath.Join(folderPathForTable, fileName), shipper_util.IsCompressedFile(fileName),
+		true, shipper_util.LoggerWithFilename(t.logger, fileName), func() (io.ReadCloser, error) {
+			return t.baseIndexSet.GetFile(ctx, t.tableName, t.userID, fileName)
+		})
+}
+
+// doConcurrentDownload downloads objects(files) concurrently. It ignores only missing file errors caused by removal of file by compaction.
+// It returns the names of the files downloaded successfully and leaves it upto the caller to open those files.
+func (t *indexSet) doConcurrentDownload(ctx context.Context, files []storage.IndexFile) ([]string, error) {
+	downloadedFiles := make([]string, 0, len(files))
+	downloadedFilesMtx := sync.Mutex{}
+
+	jobs := make([]interface{}, len(files))
+	for i := 0; i < len(files); i++ {
+		jobs[i] = i
+	}
+
+	err := concurrency.ForEach(ctx, jobs, maxDownloadConcurrency, func(ctx context.Context, job interface{}) error {
+		fileName := files[job.(int)].Name
+		err := t.downloadFileFromStorage(ctx, fileName, t.cacheLocation)
+		if err != nil {
+			if t.baseIndexSet.IsFileNotFoundErr(err) {
+				level.Info(util_log.Logger).Log("msg", fmt.Sprintf("ignoring missing file %s, possibly removed during compaction", fileName))
+				return nil
+			}
+			return err
+		}
+
+		downloadedFilesMtx.Lock()
```

Comment:
not sure you actually need a lock if you use the index instead of appending.

Reply:
yeah, that should work, but we skip adding names of the files which went missing due to compaction. This change would leave some of the elements, and the caller will have to skip the empty file names, which would not look good.